indico/htdocs/js/indico/jquery/categorypicker.js

```diff
@@ -0,0 +1,354 @@
+            var self = this;
+            var $breadcrumbs = $category.find('.breadcrumbs');
+            var availableSpace = $category.find('.title-wrapper').width();
+            var $ellipsis = $('<li>', {class: 'ellipsis'});
+            var shortened = false;
+
+            while ($breadcrumbs.outerWidth() >= availableSpace) {
+                var $segments = $breadcrumbs.children(':not(.ellipsis)');
+                var middleIndex = Math.floor($segments.length / 2);
+                if (!shortened) {
+                    $segments.eq(middleIndex).replaceWith($ellipsis);
+                } else {
+                    $segments.eq(middleIndex).remove();
+                }
+            }
+        },
+
+        _renderNavigator: function(data) {
+            var self = this;
+
+            if (data.category) {
+                self.$category.html(self._buildCurrentCategory(data.category));
+                self._ellipsizeBreadcrumbs(self.$category);
+            }
+            if (data.subcategories) {
+                _.each(data.subcategories, function(subcategory) {
+                    self.$categoryTree.append(self._buildSubcategory(subcategory));
+                });
+            }
+        },
+
+        _renderResults: function(categories) {
+            var self = this;
+
+            self.$categoryResults.html('');
+            _.each(categories, function(category) {
+                var $result = self._buildSubcategory(category, true);
+                $result.find('.icon-wrapper').append($('<i>', {
+                    class: 'icon-search',
+                    title: $T.gettext("Search result")
+                }));
+                self.$categoryResults.append($result);
+                self._ellipsizeBreadcrumbs($result);
+            });
+        },
+
+        _bindGoToCategoryOnClick: function($element, id) {
+            var self = this;
+
+            $element.on('click', function(evt) {
+                evt.preventDefault();
+                self.goToCategory(id);
+            });
+        },
+
+        _getCategoryInfo: function(id) {
+            var self = this;
+            var dfd = $.Deferred();
+
+            if (self._cache[id]) {
+                _.defer(function() {
+                    dfd.resolve(self._cache[id]);
+                });
+            } else {
+                $.ajax({
+                    url: build_url(Indico.Urls.Categories.info, {category_id: id}),
+                    dataType: 'json',
+                    error: handleAjaxError,
+                    success: function(data) {
+                        if (data) {
+                            self._cache[id] = data;
+                            dfd.resolve(data);
+                        }
+                    }
+                });
+            }
+
+            return dfd.promise();
+        },
+
+        _onAction: function(category) {
+            var self = this;
+            self.options.onAction(category);
```

Comment:
Would be nice if this could return a deferred object - in that case you'd only close the dialog once the deferred is resolved.

This could then be used during event creation to send an AJAX request to check whether event creation is actually possible in the category before closing the dialog and accepting the selection.